### ISSUE TITLE
[API Proposal]: `TextWriter.WriteLineAsync` with `CancellationToken` overloads

### ISSUE THREAD

**Issue by colejohnson66** [dotnet/runtime#106136]
### Background and motivation

Currently, the only way to use `StreamWriter.Write[Line]Async(string)` with a `CancellationToken` is to call `AsMemory` on the string, then use `StreamWriter.Write[Line]Async(ReadOnlyMemory<char>, CancellationToken)`. In fact, this is what `StreamWriter.Write[Line]Async(string)` *does*, but it uses `default(CancellationToken)` (see lines 676 and 833):

https://github.com/dotnet/runtime/blob/1d1bf92fcf43aa6981804dc53c5174445069c9e4/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L660-L685

https://github.com/dotnet/runtime/blob/1d1bf92fcf43aa6981804dc53c5174445069c9e4/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L814-L837

I propose `CancellationToken` overloads be added to `TextWriter` and have `StreamWriter` use them.

### API Proposal

```csharp
namespace System.IO;

public abstract class TextWriter
{
    // default impl would have token only affect the `Task.Factory.StartNew` invocation (same
    //   behavior as existing `Write[Line]Async(ReadOnlyMemory<char>, CancellationToken)`  methods)
    // alternatively, they could complete synchronously like `Stream` does
    public virtual Task WriteAsync(string? value, CancellationToken token);
    public virtual Task WriteLineAsync(CancellationToken token);
    public virtual Task WriteLineAsync(string? value, CancellationToken token);
}

public class StreamWriter : TextWriter
{
    // token would be passed to `WriteAsyncInternal`
    public override Task WriteAsync(string? value, CancellationToken token);
    public override Task WriteLineAsync(CancellationToken token);
    public override Task WriteLineAsync(string? value, CancellationToken token);
}
```

### API Usage

Old code:

```csharp
await using StreamWriter writer = new(...);
await writer.WriteLine(string.Create(CultureInfo.InvariantCulture, $"...").AsMemory(), token);
```

New code:

```csharp
await using StreamWriter writer = new(...);
await writer.WriteLine(string.Create(CultureInfo.InvariantCulture, $"..."), token);
```

### Alternative Designs

An analyzer to turn `StreamWriter.Write[Line]Async(string)` into `StreamWriter.Write[Line]Async(ReadOnlyMemory<char>, CancellationToken)`

### Risks

Code previously using `TextWriter.WriteLineAsync()` and `TextWriter.Write[Line]Async(string)` with a `CancellationToken` in scope will get new warnings to pass in that token.

### Open Questions

* Should the cancellation token have a default value (i.e., `CancellationToken token = default` in the signatures)?

**Comment by colejohnson66** [comment]
Yes. `WriteAsync(CancellationToken)` was a copy/paste bug. Such an operation wouldn't make sense. They also should've been virtual.

Updated proposal.

**Comment by mio991** [comment]
I stumbled over the same problem. I thought about giving it a shot, would this be a good first issue?

**Comment by stephentoub** [comment]
> I stumbled over the same problem. I thought about giving it a shot, would this be a good first issue?

Thanks for the interest. We need to review the proposal first.

**Comment by bartonjs** [comment]
[Video](https://www.youtube.com/watch?v=p857nTg9lTg&t=0h32m40s)

* Approved as non-virtual (they can defer to the ReadOnlyMemory implementation).  If we need virtual later we can do that.

```csharp
namespace System.IO;

public abstract class TextWriter
{
    public Task WriteAsync(string? value, CancellationToken token);
    public Task WriteLineAsync(CancellationToken token);
    public Task WriteLineAsync(string? value, CancellationToken token);
}
```
